Sources/MembraneRTC/Media/SoundDetection.swift

```diff
@@ -0,0 +1,179 @@
+    let mainMixerNode = engine.mainMixerNode
+    let mixerFormat = AVAudioFormat(
+      commonFormat: .pcmFormatFloat32, sampleRate: inputFormat.sampleRate, channels: 1,
+      interleaved: false)
+    engine.connect(mixerNode, to: mainMixerNode, format: mixerFormat)
+  }
+
+  /// Starts the sound detection process with the specified volume threshold.
+  ///
+  /// - Parameter volumeThreshold: The threshold value in decibels (dB) above which a sound is considered detected.
+  public func start(_ volumeThreshold: Int = 60) {
+    let tapNode: AVAudioNode = mixerNode
+    let format = tapNode.outputFormat(forBus: 0)
+
+    tapNode.installTap(
+      onBus: 0, bufferSize: 4096, format: format,
+      block: {
+        (buffer, time) in
+        self.processAudioBufferForSoundDetection(from: buffer, volumeThreshold)
+      })
+
+    do {
+      try engine.start()
+      isRecording = true
+    } catch let error {
+      print("Failed to start audio engine: \(error.localizedDescription)")
+    }
+  }
+
+  /// Processes the provided AVAudioPCMBuffer for sound detection.
+  ///
+  /// - Parameters:
+  ///   - buffer: The audio buffer.
+  ///   - volumeThreshold: The threshold value in decibels (dB) above which a sound is considered detected.
+  fileprivate func processAudioBufferForSoundDetection(
+    from buffer: AVAudioPCMBuffer, _ volumeThreshold: Int = 60
+  ) {
+    let amplitude = getMaxAmplitude(from: buffer)
+    let soundVolume = calculateValue(from: amplitude)
+    onVolumeChangedListener?(soundVolume)
+    onSoundDetectedListener?(soundVolume > volumeThreshold)
+  }
+
+  /// Stops the sound detection process.
+  public func stop() {
+    mixerNode.removeTap(onBus: 0)
+    engine.stop()
+    isRecording = false
+  }
+
+  /// Method to calculate the maximum amplitude in the buffer.
+  ///
+  /// - Parameter buffer: The audio buffer.
+  /// - Returns: The maximum amplitude value from the buffer.
+  fileprivate func getMaxAmplitude(from buffer: AVAudioPCMBuffer) -> Int {
+    let convertedBuffer = convertFloat32To16(buffer: buffer)
+    guard let data = convertedBuffer?.int16ChannelData else {
+      return 0
+    }
+    let frameLength = Int(buffer.frameLength)
+    var maxAmplitude: Int16 = 0
+    for i in 0..<frameLength {
+      maxAmplitude = max(data[0][i], maxAmplitude)
+    }
+    return Int(maxAmplitude)
+  }
+
+  /// Converts PCM Buffer from Float32 to Int16
+  ///
+  /// - Parameter buffer: The audio buffer in Float32.
+  /// - Returns: The converted audio buffer in Int16 format, returned nil if conversion failed.
+  fileprivate func convertFloat32To16(buffer: AVAudioPCMBuffer) -> AVAudioPCMBuffer? {
+    let sourceFormat = buffer.format
+    let targetFormat = AVAudioFormat(
+      commonFormat: .pcmFormatInt16, sampleRate: sourceFormat.sampleRate,
+      channels: sourceFormat.channelCount,
+      interleaved: true)
+
+    guard let converter = AVAudioConverter(from: sourceFormat, to: targetFormat!) else {
```

Comment:
I advise against using `!` operator, it caused a few crashes in this project already, you can put it in the guard statement too (you can have multiple variables in one guard)